### PR TITLE
Add a service socket

### PR DESCRIPTION
**Pull request by Xaekai**
to enable out of band access to the process commandline

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -75,5 +75,33 @@ function handleLine(line) {
...
+        // simple validation for IPv4
+        // TODO: IPv6
+        if(args.length === 2 && args[0].match(/^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/)){
+            require("./lib/database").globalBanIP(args[0], args[1], function (err, res) {
+                if (!err) {
+                    Logger.eventlog.log("[acp] " + "SYSTEM" + " global banned " + args[0]);
+                }
+            })
+        }
+    } else if (line.indexOf("/clearban") === 0) {
```
Comment (calzoneman) [inline index.js]: /unglobalban or something like that would be more consistent naming.

Replace the IP address regex with `net.isIPv4()`: https://nodejs.org/api/net.html#net_net_isipv4_input

File: index.js [hunk 2]
```diff
@@ -75,5 +75,33 @@ function handleLine(line) {
...
+                if (!err) {
+                    Logger.eventlog.log("[acp] " + "SYSTEM" + " un-global banned " + args[0]);
+                }
+            })
+        }
     }
 }
+
+// Go Go Gadget Service Socket
+if(Config.get("service-socket.enabled")){
```
Comment (calzoneman) [inline index.js]: Spacing around  `if () {`

File: index.js [hunk 3]
```diff
@@ -75,5 +75,33 @@ function handleLine(line) {
...
+    Logger.syslog.log("Opening service socket");
+    var ServiceSocket = require('./lib/servsock');
+    var server = new ServiceSocket;
```
Comment (calzoneman) [inline index.js:119]: Is this even legal in ECMAScript?

File: servcmd.sh.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env node
+/*
+**  CyTube Service Socket Commandline
+*/
+
+var Config = require("./lib/config");
+Config.load("config.yaml");
+
+if(!Config.get("service-socket.enabled")){
+	console.error('The Service Socket is not enabled.');
```
Comment (calzoneman) [inline servcmd.sh.js]: INDENTING WITH TABS REEEEEEEEEEEEEEEE

File: servcmd.sh.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        if(msg === '__disconnect'){
+            console.log('Server shutting down. Terminating.')
+            return cleanup();
+        }
+
+        // Generic message handler
+        console.info('Server:', data)
+    })
+    .on('error', (data)=>{
```
Comment (calzoneman) [inline servcmd.sh.js]: Spacing around arrow operator

File: src/servsock.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                }
+                return this.openServiceSocket();
+            });
+        });
+    }
+
+    openServiceSocket(){
+        this.server = net.createServer((stream)=>{
+            let self = Date.now();
+            this.connections[self] = (stream);
```
Comment (calzoneman) [inline src/servsock.js]: Why is `stream` wrapped in parens

File: src/servsock.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                    console.error(err); process.exit(0);
+                }
+                return this.openServiceSocket();
+            });
+        });
+    }
+
+    openServiceSocket(){
+        this.server = net.createServer((stream)=>{
+            let self = Date.now();
```
Comment (calzoneman) [inline src/servsock.js]: Throughout CyTube, `self` is traditionally used as a lexical closure around `this` for legacy ECMAScript 5 code where JavaScript would mangle `this` in callbacks/nested functions.  Using it as something unrelated here seems like a weird naming choice.

File: src/servsock.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            this.connections[self] = (stream);
+            stream.on('end', ()=>{
+                delete this.connections[self];
+            });
+            stream.on('data', (msg)=>{
+                this.handler(msg.toString());
+            });
+        })
+        .listen(this.socket);
```
Comment (calzoneman) [inline src/servsock.js]: Convention throughout CyTube is to have chained calls start on the same line

File: src/servsock.js (hunk 6 above)
Comment (Xaekai) [inline src/servsock.js]: Vestigial.

**Comment by Xaekai** [comment]
Updated.

File: index.js [hunk 9]
```diff
@@ -75,5 +76,31 @@ function handleLine(line) {
         }
     } else if (line.indexOf("/reload-partitions") === 0) {
         sv.reloadPartitionMap();
+    } else if (line.indexOf("/globalban") === 0) {
+        var args = line.split(" "); args.shift();
+        if (args.length === 2 && net.isIPv4(args[0]) || net.isIPv6(args[0])) {
```
Comment (calzoneman) [inline index.js]: Operator precedence means that this is parsed as `(args.length === 2 && net.isIPv4(args[0])) || net.isIPv6(args[0])`, also if you want to support both you can just check if `net.isIP()` is nonzero.

**Comment by Xaekai** [comment]
Updated and retested.

File: index.js [hunk 10]
```diff
@@ -45,6 +45,7 @@ process.stdin.on("data", function (data) {
     }
 });
 
+var net = require('net');
```
Comment (calzoneman) [inline index.js]: Remove unused import

File: src/servsock.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
+var fs = require('fs');
+var net = require('net');
+
+export default class ServiceSocket {
```
Comment (calzoneman) [inline src/servsock.js:4]: Does this class need error event handlers?  In case the client connection breaks or something.

File: index.js (hunk 10 above)
Comment (Xaekai) [inline index.js]: It's not unused. It's used to to validate user input, which was your suggestion over a regex.

Edit: I can change it to just import the one function though.

File: src/servsock.js (hunk 11 above)
Comment (Xaekai) [inline src/servsock.js:4]: I have thus far been unable to break anything. Sending SIGKILL to the client leaves orphaned children (the tail on the log files) but doesn't affect the service at all. Sending SIGTERM to the service after SIGKILL'ing a client shuts down the service with no errors.

**Pull request merged**